include/pybind11/gil.h

```diff
@@ -199,3 +199,43 @@ class gil_scoped_release {
 PYBIND11_NAMESPACE_END(PYBIND11_NAMESPACE)
 
 #endif // !PYBIND11_SIMPLE_GIL_MANAGEMENT
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+class scoped_critical_section {
+public:
+    /// This does not do anything if there's a GIL. On free-threaded Python,
+    /// it locks an object. This uses the CPython API, which has limits
+    scoped_critical_section(handle obj) : single(true) {
+#ifdef Py_GIL_DISABLED
+        PyCriticalSection_Begin(&section, obj.ptr());
+#endif
+    }
+
+    scoped_critical_section(handle obj1, handle obj2) : single(true) {
```

Comment:
`single(false)`
?
But, maybe `has2` instead as variable name? (Seems slightly more intuitive to me.)

Reply:
I originally had `dual`, but didn't like the most common one being false, seemed inverted (which is why this is typo is here).